File: javascript/rest-api/rest-api.js
export class RestAPI {

  #db;
  constructor(caloteiros) {
    this.#db = { users: [] };
    caloteiros.users.forEach(user => this.#db.users.push(user));
  }

  get(url) {
    let request = url.split('?');

    if (request.length === 1) {
      return this.#db;
    }

    let searchedUser = request[1].split('=')[1];
    let filteredUsers = [
      ...
      this.#db.users.filter(user => user.name === searchedUser)
    ];

    return { users: filteredUsers };
  }

  post(url, payload) {

    switch(url) {
      case '/add':
        return this.addUser(payload);
      case '/iou':
        return this.iou(payload);
    }
  }

  addUser(userToAdd) {
    let u = {
      name: userToAdd.user,
      owes: {},
      owed_by: {},
      balance: 0,
    };

    this.#db.users.push(u);

    return u;
  }

  iou(lendersBorrowers) {
    let { lender, borrower, amount } = lendersBorrowers;

    let l = this.#db.users.find(user => user.name === lender);
    let b = this.#db.users.find(user => user.name === borrower);

    if (l.owed_by[borrower]) {
      l.owed_by[borrower] = l.owed_by[borrower] + amount;
      b.owes[lender] = b.owes[lender] + amount;
    } else if (l.owes[borrower]) {
      let debt = l.owes[borrower] - amount;
      if (debt > 0) {
        l.owes[borrower] = debt;
        b.owed_by[lender] = debt;
      } else {
        delete l.owes[borrower];
        delete b.owed_by[lender];

        if (debt < 0) {
          l.owed_by[borrower] = Math.abs(debt);
          b.owes[lender] = Math.abs(debt);
        }
      }
    } else {
      l.owed_by[borrower] = amount;
      b.owes[lender] = amount;
    }
    l.balance = l.balance + amount;
    b.balance = b.balance - amount;


    // Tem que ordernar essa merda senão o teste não 
    // passa mesmo com os objetos contendo os dados corretos.
    // Tomanocú viu...
    let u = [l, b];
    u.sort((a, c) => (a.name > c.name) ? 1 : ((c.name > a.name) ? -1 : 0));
    let ret = { users: u };

    return ret;
  }
}

File: java/rest-api/src/main/java/RestApi.java
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class RestApi{

    private final List<User> userList;

    public RestApi(User ...users) {
        this.userList = new ArrayList<>();
        this.userList.addAll(List.of(users));
    }

    public String get(String url, JSONObject ...payload) {

        JSONObject jsonObject = new JSONObject();

        if (payload.length > 0) {
            String name = ((JSONArray)payload[0].get("users")).getString(0);
            User searchedUser = this.userList.stream()
                    .filter(user -> user.name().equals(name))
                    .collect(Collectors.toList()).get(0);

            JSONArray jsonArray = new JSONArray();
            jsonArray.put(userToJSON(searchedUser));
            jsonObject.put("users", jsonArray);

        } else {
            jsonObject.put("users", this.userList);
        }

        System.out.println(jsonObject);

        return jsonObject.toString();
    }

    public String post(String url, JSONObject payload) {

        if (url.equals("/add")) {
            String name = (String) payload.get("user");

            User.Builder builder = User.builder();
            builder.setName(name);
            User user = builder.build();

            this.userList.add(user);
            return userToJSON(user).toString();
        }

        if (url.equals("/iou")) {
           return addIou(payload);
        }
        return null;
    }

    private String addIou(JSONObject payload) {
        String lenderName = (String) payload.get("lender");
        String borrowerName = (String) payload.get("borrower");
        double amount = (double) payload.get("amount");
        JSONArray jsonArray = new JSONArray();

        boolean usersAlreadyInDB = this.userList.stream()
                .anyMatch(user -> user.name().equals(lenderName));


        if (usersAlreadyInDB) {
            for (int i = 0; i < this.userList.size(); i++) {
                User currentUser = this.userList.get(i);
               if (currentUser.name().equals(lenderName)) {
                   List<Iou> owes = new ArrayList<>(currentUser.owes());
                   List<Iou> owedBy = new ArrayList<>(currentUser.owedBy());
                   User.Builder builder = User.builder();
                   builder.setName(lenderName);
                   owes.forEach(iou -> builder.owes(iou.name, iou.amount));
                   owedBy.forEach(iou -> builder.owedBy(iou.name, iou.amount));
                   builder.owedBy(borrowerName, amount);
                   User lender = builder.build();
                   this.userList.remove(i);
                   this.userList.add(i, lender);
                   jsonArray.put(userToJSON(lender));
                   System.out.println("Lender added: " + jsonArray);
               } else if (currentUser.name().equals(borrowerName)) {
                   List<Iou> owes = new ArrayList<>(currentUser.owes());
                   List<Iou> owedBy = new ArrayList<>(currentUser.owedBy());
                   User.Builder builder = User.builder();
                   builder.setName(borrowerName);
                   owes.forEach(iou -> builder.owes(iou.name, iou.amount));
                   owedBy.forEach(iou -> builder.owedBy(iou.name, iou.amount));
                   builder.owes(lenderName, amount);
                   User borrower = builder.build();
                   this.userList.remove(i);
                   this.userList.add(i, borrower);
                   jsonArray.put(userToJSON(borrower));
               }
            }

        } else {
            System.out.println("Aqui||");
            User.Builder builder = User.builder();
            builder.setName(lenderName);
            builder.owedBy(borrowerName, amount);
            User lender = builder.build();

            builder = User.builder();
            builder.setName(borrowerName);
            builder.owes(lenderName, amount);
            User borrower = builder.build();

            this.userList.add(lender);
            this.userList.add(borrower);
            jsonArray.put(userToJSON(lender));
            jsonArray.put(userToJSON(borrower));
        }

        JSONObject jo = new JSONObject()
                .put("users", jsonArray);

        System.out.println(jo);

        return jo.toString();
    }

    private JSONObject userToJSON(User user) {
        String n = user.name();
        List<Iou> o = user.owes();
        List<Iou> oBy = user.owedBy();
        double balance = oBy.stream().map(iou -> iou.amount)
                .mapToDouble(Double::doubleValue).sum() -
                o.stream().map(iou -> iou.amount)
                        .mapToDouble(Double::doubleValue).sum();

        JSONObject owes = new JSONObject();
        o.forEach(iou -> owes.put(iou.name, iou.amount));

        JSONObject owedBy = new JSONObject();
        oBy.forEach(iou -> owedBy.put(iou.name, iou.amount));

        return new JSONObject()
                .put("name", n)
                .put("owes", owes)
                .put("owedBy", owedBy)
                .put("balance", balance);
    }
}